File: day008_string_traversal/primary_odd_even_index/solution.cpp
#include <iostream>
#include <string>
#include <sstream>

using namespace std;

// O(N) Time, O(N) Space - Optimal (Single Pass)
void solve_odd_even_index() {
    string S;
    // Read the entire line as the string
    if (!getline(cin, S)) return;

    string even_chars = "";
    string odd_chars = "";

    // Single Linear Pass (O(N))
    for (int i = 0; i < S.length(); ++i) {
        if (i % 2 == 0) {
            even_chars += S[i];
        } else {
            odd_chars += S[i];
        }
    }
    
    cout << even_chars << endl;
    cout << odd_chars << endl;
}

// O(N) Time, O(N) Space - Brute (Same fundamental approach)
void solve_odd_even_index_brute() {
    solve_odd_even_index(); 
}

int main() {
    solve_odd_even_index();
    return 0;
}

File: day008_string_traversal/primary_odd_even_index/solution.py
import sys

def solve_odd_even_index():
    try:
        S = sys.stdin.readline().strip()
    except:
        return

    # Optimized: O(N) Time, O(N) Space - Single Pass
    even_chars = []
    odd_chars = []
    
    for i, char in enumerate(S):
        if i % 2 == 0:
            even_chars.append(char)
        else:
            odd_chars.append(char)
            
    # Print the two resulting strings
    print("".join(even_chars))
    print("".join(odd_chars))

# Brute Force: Same O(N) logic, perhaps using two separate loops (less efficient constant factor)
def solve_odd_even_index_brute():
    # Since the optimal logic is a single O(N) pass, the "brute" method would be very similar.
    solve_odd_even_index()

if __name__ == "__main__":
    solve_odd_even_index()